modules/testkit/src/main/scala/lucuma/core/model/arb/ArbProposalClass.scala

```diff
@@ -0,0 +1,141 @@
+
+  implicit val arbExchange: Arbitrary[Exchange] =
+    Arbitrary(arbitrary[IntPercent].map(Exchange(_)))
+
+  implicit val cogExchange: Cogen[Exchange] =
+    Cogen[IntPercent].contramap(_.minPercentTime)
+
+  implicit val arbFastTurnaround: Arbitrary[FastTurnaround] =
+    Arbitrary(arbitrary[IntPercent].map(FastTurnaround(_)))
+
+  implicit val cogFastTurnaround: Cogen[FastTurnaround] =
+    Cogen[IntPercent].contramap(_.minPercentTime)
+
+  implicit val arbPoorWeather: Arbitrary[PoorWeather] =
+    Arbitrary(arbitrary[IntPercent].map(PoorWeather(_)))
+
+  implicit val cogPoorWeather: Cogen[PoorWeather] =
+    Cogen[IntPercent].contramap(_.minPercentTime)
+
+  implicit val arbQueue: Arbitrary[Queue] =
+    Arbitrary(arbitrary[IntPercent].map(Queue(_)))
+
+  implicit val cogQueue: Cogen[Queue] =
+    Cogen[IntPercent].contramap(_.minPercentTime)
+
+  implicit val arbSystemVerification: Arbitrary[SystemVerification] =
+    Arbitrary(arbitrary[IntPercent].map(SystemVerification(_)))
+
+  implicit val cogSystemVerification: Cogen[SystemVerification] =
+    Cogen[IntPercent].contramap(_.minPercentTime)
+
+  implicit val arbLargeProgram: Arbitrary[LargeProgram] =
+    Arbitrary {
+      for {
+        minPct    <- arbitrary[IntPercent]
+        minTotPct <- arbitrary[IntPercent]
+        totalTime <- arbitrary[Duration]
+      } yield LargeProgram(minPct, minTotPct, totalTime)
+    }
+
+  implicit val cogLargeProgram: Cogen[LargeProgram] =
+    Cogen[(IntPercent, IntPercent, Duration)].contramap(p =>
+      (p.minPercentTime, p.minPercentTotalTime, p.totalTime)
+    )
+
+  implicit val arbIntensive: Arbitrary[Intensive] =
+    Arbitrary {
+      for {
+        minPct    <- arbitrary[IntPercent]
+        minTotPct <- arbitrary[IntPercent]
+        totalTime <- arbitrary[Duration]
+      } yield Intensive(minPct, minTotPct, totalTime)
+    }
+
+  implicit val cogIntensive: Cogen[Intensive] =
+    Cogen[(IntPercent, IntPercent, Duration)].contramap(p =>
+      (p.minPercentTime, p.minPercentTotalTime, p.totalTime)
+    )
+
+  implicit val arbProposalClass: Arbitrary[ProposalClass] =
+    Arbitrary(
+      Gen.oneOf(
+        arbitrary[Classical],
+        arbitrary[DemoScience],
+        arbitrary[DirectorsTime],
+        arbitrary[Exchange],
+        arbitrary[FastTurnaround],
+        arbitrary[PoorWeather],
+        arbitrary[Queue],
+        arbitrary[SystemVerification],
+        arbitrary[LargeProgram],
+        arbitrary[Intensive]
+      )
+    )
+
+  implicit val cogProposalClass: Cogen[ProposalClass] =
+    Cogen[
+      (
+        Option[Classical],
+        Option[DemoScience],
+        Option[DirectorsTime],
+        Option[Exchange],
+        Option[FastTurnaround],
+        Option[PoorWeather],
+        Option[Queue],
+        Option[SystemVerification],
+        Option[LargeProgram],
+        Option[Intensive]
+      )
+    ].contramap(p =>
+      (
+        ProposalClass.classical.getOption(p),
```

Comment:
This looks nicer than going with nested `Either`s.

I havent thought this fully through, bu I'm a bit worried that it may present a chicken and egg issue: is this `Cogen` used to test the `Prism`s that are being used to derive it?

Reply:
Also, would it be an issue if the prisms were ill formed such that two or more of them returned a value for the same `p`? (Just asking, I don't have an answer).

Reply:
Okay... I did a little research and testing. Cogen is used for seed generation, so I don't think nested Either vs tuple of Option makes a difference.

I removed the `Cogen[ProposalClass]` (and by necessity the `Cogen[Proposal]`. The only tests affected by the removals were `EqTests[ProposalClass]`, ˙LensTests(Proposal.proposalClass)' and `EqTests[Proposal]`. The first 2 were due to the lack of `Cogen[ProposalClass]`. The prism tests don't involve the `Cogen[ProposalClass]`, so it seems safe to use them here. 

Any objections?

Reply:
> Any objections?

No objections here, regardless of how it goes.  It seemed like it might be possible to avoid the `.asRight` chains but in reality it's a minor issue.